app/src/editor/canvas/components/Ports/Port/index.jsx

```diff
@@ -0,0 +1,178 @@
+        setOptions(port.id, {
+            displayName,
+        })
+    }, [port.id, setOptions])
+
+    const onOptionToggle = useCallback((key) => {
+        setOptions(port.id, {
+            [key]: !port[key],
+        })
+    }, [port, setOptions])
+
+    const onValueChange = useCallback((value: any) => {
+        onValueChangeProp(port.id, value)
+    }, [port.id, onValueChangeProp])
+
+    const { isDragging, data } = useContext(DragDropContext)
+    const { portId } = data || {}
+    const dragInProgress = !!isDragging && portId != null
+
+    useEffect(() => {
+        window.addEventListener('mousedown', onWindowMouseDown)
+        window.addEventListener('keydown', onGlobalKeyDown)
+        window.addEventListener('blur', onWindowBlur)
+
+        return () => {
+            window.removeEventListener('mousedown', onWindowMouseDown)
+            window.removeEventListener('keydown', onGlobalKeyDown)
+            window.removeEventListener('blur', onWindowBlur)
+        }
+    }, [onWindowMouseDown, onGlobalKeyDown, onWindowBlur])
+
+    return (
+        <div
+            className={cx(styles.root, {
+                [styles.dragInProgress]: !!dragInProgress,
+            })}
+        >
+            {contextMenuTarget && (
+                <Menu
+                    api={api}
+                    dismiss={dismiss}
+                    port={port}
+                    setPortOptions={setOptions}
+                    target={contextMenuTarget}
+                />
+            )}
+            {port.canToggleDrivingInput && (
+                <Option
+                    activated={!!port.drivingInput}
+                    className={styles.portOption}
+                    disabled={!!isRunning}
+                    name="drivingInput"
+                    onToggle={onOptionToggle}
+                />
+            )}
+            {!isInput ? (
+                <Cell className={styles.spaceholder} />
+            ) : plug}
+            <Cell>
+                <EditableText
+                    disabled={!!isRunning}
+                    editing={editingName}
+                    onChange={onNameChange}
+                    setEditing={setEditingName}
+                >
+                    {port.displayName || startCase(port.name)}
+                </EditableText>
+            </Cell>
+            {hasInputField && (
+                <Cell>
+                    <Value
+                        canvas={canvas}
+                        port={port}
+                        onChange={onValueChange}
+                    />
+                </Cell>
+            )}
+            {!isInput && plug}
+            {port.canBeNoRepeat && (
+                <Option
+                    activated={!!port.noRepeat}
+                    className={styles.portOption}
+                    disabled={!!isRunning}
+                    name="noRepeat"
+                    onToggle={onOptionToggle}
+                />
+            )}
```

Comment:
Would be good to make sure all of these `conditional && <Component />` were forcibly coerced to booleans i.e. `!!conditional && <Component />` so we can't possibly accidentally render a `0`, ever. Figuring out exactly which code path happened to generate a zero can be very time consuming. I know `flow` should prevent this, but `flow` isn't perfect and the cost to prevent this from ever happening is just adding `!!` so it seems to me a low cost/high yield habit.